Test platform mingw - For discussion

hamishcoleman opened the pull request:
This updates the automated tests to run on windows/mingw.

Which has required a number of changes
- autoconf needed to be replaced for this build
- none of the configure options are available
- the Makefiles have been expanded and overhauled to add support for win32

Most importantly, while the build completes, the created library doesnt create the same transform output as a version built on ubuntu.  Which makes the build fail the tests.

I have also documented a simple set of steps to make the same build locally and this branch along with those instructions can be used to assist in debugging the issue with the transform.

hamishcoleman committed:
Provide a minimal reimplementation of our autoconf, to try windows builds

hamishcoleman committed:
Try building with windows

hamishcoleman committed:
Fix thinko in spelling

hamishcoleman committed:
Ensure shell script runs inside a shell

hamishcoleman committed:
Add a hack to aid include discovery

hamishcoleman committed:
Just keep adding tech debt...

hamishcoleman committed:
Assume that we will have slashes in some of the replacement strings and avoid that char with sed

hamishcoleman committed:
Restore one slash

hamishcoleman committed:
Hack around the tools makefile interdependancy bug

hamishcoleman committed:
A correct cflags include hack for each compile dir

hamishcoleman committed:
Ensure we link against winsock (note, even though this says 32bit, it should link the 64bit library ... I think)

hamishcoleman committed:
Bad link ordering if we dont use LDLIBS

hamishcoleman committed:
Remove unused make variable

hamishcoleman committed:
Remove makefile duplication using inheritance (this does mean you can no longer cd tools; make, but must do make tools)

hamishcoleman committed:
Add missing library for win32

hamishcoleman committed:
Show OS variable

hamishcoleman committed:
Make hack autoconf more robust for tests on non gitlab runners

hamishcoleman committed:
Remove no longer used substitutions from hack autoconf

hamishcoleman committed:
Add missing include path to tools under win32

hamishcoleman committed:
Build the win32 subdir when the compiler is Msys

hamishcoleman committed:
The different subdirs have different dependancies

hamishcoleman committed:
Ensure we can find the include files

hamishcoleman committed:
Fix library link ordering

hamishcoleman committed:
Ensure the tools dir can find the special win32 lib

hamishcoleman committed:
Deal with the differing basic type sizes on both linux/64bit and windows/64bit

hamishcoleman committed:
Document the steps to mimic the github windows/mingw build locally - to allow for simpler debugging

hamishcoleman commented:
@Logan007 hopefully, this appeases your curiosity

Logan007 commented:
Indeed, it does! And to some extent, I feel urged to jump right into MinGW debugging...

hamishcoleman commented:
jump in - thats what I was hoping to inspire!

(and I'm not so good on debugging crypto miscompiles)

Logan007 commented:
Still fighting with mingw compilation...

I had to extend the `-I` to `./win32` and `.` in Makefile.

But then, I see a lot of `undefined reference to` ... network related stuff such as __imp_ntohs and others...

hamishcoleman commented:
Strange, those are all things that I fixed already.  Can you share your build log?

hamishcoleman committed:
Ensure branch name in instructions matches my test branch name

hamishcoleman commented:
It should not have made any difference for your testing, but the branch name in my instructions was an old branch name.

Logan007 commented:
Please see the [log](https://github.com/ntop/n2n/files/7253032/log.txt).

hamishcoleman commented:
Well, I'm running all my steps from a bash shell.  Which probably means that any of the shell commands that the Makefile run dont work - which I think has meant that the Makefile didnt realise that it was running on windows/mingw

Logan007 commented:
I ran the script on bash (Linux) as well, removed the ^M and copied back to Windows...

Do you use bash on Windows?

hamishcoleman commented:
I used bash on Windows, yes.  The Makefile needs to run with a bash shell as well as the hack_fakeautoconf script.

From the desktop, I could start/run/bash and it started, or from a windows ssh session, I could run "C:\Program Files\Git\usr\bin\bash"

hamishcoleman committed:
Clarify the shell needed to build with mingw

hamishcoleman committed:
Since the makefile depends on knowing the OS, raise a fatal error if we cannot determine this

Logan007 commented:
Meanwhile... I have been able to get it compiled. Thanks for setting me up! Will be gone for a while...debugging :wink:

hamishcoleman commented:
Good Luck!

Let me know if you come up against anything else in the buildsystem or docs that might trap people.

Logan007 commented:
Interesting (?) observation: I did not use choco but `sudo apt-get install gcc-mingw-w64-i686` (Ubuntu subsytem installed).

The tests do not complain...

hamishcoleman commented:
The Makefile system is hardcoded to use regular gcc, so cross-compiling from a linux container will just use the containers system compiler.

Logan007 commented:
I think I found the bug! (#830) Hopefully...

Without your help, I wouldn't even have gotten a working MinGW compile. Again, thank you!

I guess this PR needs some changes before merging?

hamishcoleman commented:
It is not ideal as it is, no (though it was surprisingly easy to rip out the autotools) - it is probably better to add easier cross compiling support to the Makefiles, and then using gcc-mingw-w64-i686

hamishcoleman committed:
Handling different compile environments is hard.

- Linux: sane and reasonable results for both uname -s (=Linux) and
  uname -o (=GNU/Linux)
- Windows/Mingw: insane results for uname -s
  (=MSYS_NT-$MAJOR.$MINOR-$BUILDNR) but sane results for uname -o (Msys)
- Macos: sane results for uname -s (=Darwin) but does not support
  uname -o at all

hamishcoleman committed:
Revamp the way that Mingw is detected

hamishcoleman committed:
Avoid attempting to generate gcovr report when running under windows

hamishcoleman committed:
Whoops, isolate the right step

hamishcoleman commented:
Yey, it can pass the tests automatically!

Logan007 commented:
So, shall we merge?

hamishcoleman commented:
I want to look closer at the diff and see if I can make it less invasive.

I've also been trying to work out how to make mingw on ubuntu compile the same code, so I'm looking at making the compile time target environment detection a little more generic and reusable - which might help justify some of the other changes in here.

Logan007 commented:
Okay, just let me know.

By the way, I think the new compile instructions somehow got placed in the middle of some other section...

hamishcoleman commented:
ha, I think I threw them anywhere in a mild hurry.  They need looking at anyway - now you have added your patch, they need to stop saying "do not use" :-)

hamishcoleman committed:
Fix spelling mistake

hamishcoleman committed:
win32/Makefile: Remove unused setting and add comment

hamishcoleman committed:
ensure that all win32 includes use the same expected path

hamishcoleman committed:
Allow simpler cross compilation by letting configure pass the CC and AR environment through

hamishcoleman committed:
Avoid multiple '_CRT_SECURE_NO_WARNINGS redefined' warnings

hamishcoleman committed:
Convert to a consolidated CONFIG_TARGET variable to select any different compile options

hamishcoleman committed:
Use the more generic printf defines to avoid warnings on mingw

hamishcoleman committed:
Update mingw build docs

hamishcoleman committed:
English better for reader happy make

hamishcoleman commented:
@Logan007 I am happier with this branch now.

There is one change that might trip some people up - with these patches applied, you can no longer "cd tools; make" but must run the make from the top dir.

hamishcoleman committed:
Address a number of mingw compiler warnings

hamishcoleman committed:
Fix Visual C compile

hamishcoleman committed:
Be sure to document some of the hacky nature of the mingw build

Logan007 commented:
Thank you very much for this great contribution!
